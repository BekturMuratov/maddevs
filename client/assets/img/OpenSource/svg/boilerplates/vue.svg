<svg xmlns="http://www.w3.org/2000/svg" width="336" height="125" fill="none" xmlns:v="https://vecta.io/nano"><g clip-path="url(#A)"><path d="M206.784.001l-38.797 67.203L129.191.001H-.008l167.996 290.997L335.984.001h-129.2z" fill="#41b883"/><path d="M206.785.001l-38.797 67.203L129.191.001h-62l100.798 174.595L268.786.001h-62.001z" fill="#34495e"/></g><defs><clipPath id="A"><path fill="#fff" d="M0 0h336v291H0z"/></clipPath></defs></svg>