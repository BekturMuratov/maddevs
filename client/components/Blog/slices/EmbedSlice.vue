<template>
  <div class="embed" :class="slice.items[0].embed.type" v-if="slice.items.length">
    <prismic-embed :field="slice.items[0].embed" target="_blank" :class="`embed__${slice.items[0].embed.type}`"/>
    <template v-if="slice.items[0].embed.type === 'link'">
      <div class="embed__image-wrapper">
        <div class="embed__image" :style="{backgroundImage: `url(${slice.items[0].embed.thumbnail_url})`}" />
      </div>
    </template>
  </div>
</template>

<script>
export default {
  props: {
    slice: {
      type: Object,
      default: null
    }
  },
  name: 'embed-slice',
  created() {
    if (!this.slice.items.length) {
      return;
    }

    if (this.slice.items[0].embed.html) {
      this.slice.items[0].embed.html = this.slice.items[0].embed.html
        .replace('<h1>', '<div class="embed__title">')
        .replace('</h1>', '</div>')
        .replace(/<img[^>]*>/g, '')
        .replace(/<a href="http[^"]*"/, match => `${match} target="_blank"`);
    }

    if (this.slice.items[0].embed.type === 'video') {
      this.slice.items[0].embed.html = this.slice.items[0].embed.html.replace(/height="[0-9]*"/, 'height="500"').replace(/width="[0-9]*"/, 'width="100%"');
    }
  }
};
</script>

<style lang="scss" scoped>
  @import '../../../assets/styles/_vars';
  .embed {

    /deep/ iframe {
      max-width: 100%;
    }

    &.link {
      display: flex;
      margin: 25px 0;
      border: 1px solid $border-color--silver;
    }

    /deep/ > div {
      a {
        text-decoration: none;
      }

      img {
        display: none;
      }

      h1,
      p {
        letter-spacing: -0.02em;
        line-height: 129%;
      }

      p {
        font-size: 13px;
        font-family: 'Poppins-Regular', sans-serif;
        color: $text-color--grey-pale;
      }
    }

    &__link {
      display: flex;
      padding: 24px 0 24px 24px;
    }

    /deep/ &__title {
      margin-top: 0;
      margin-bottom: 6px;
      font-size: 21px;
      font-family: 'Poppins-Bold', sans-serif;
      font-weight: 600;
      color: $text-color--black-cases;
      line-height: 27.34px;
    }

    &__image {
      width: 100%;
      height: 100%;
      min-width: 136px;
      background-size: cover;
      background-position: center;
      background-repeat: no-repeat;

      &-wrapper {
        width: 33%;
        padding: 24px !important;
      }
    }
  }

  @media screen and (max-width: 1024px) {
    .embed {
      &.link {
        flex-wrap: wrap;
      }

      &__link {
        order: 2;
        width: 100%;
        padding-right: 24px;
      }
      &__image {
        height: 40vw;

        &-wrapper {
          order: 1;
          width: 100%;
        }
      }
    }
  }
</style>
