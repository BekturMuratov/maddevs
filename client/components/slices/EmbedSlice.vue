<template>
  <Component
    :is="isTag.is"
    v-if="item"
    v-bind="isTag"
  >
    <PrismicEmbed
      :field="embedFieldData"
      :class="`embed__${item.embed.type}-content`"
    />
    <template v-if="item.embed.type === 'link'">
      <div class="embed__image-wrapper">
        <img
          v-lazy-load
          :data-src="embedImage.url"
          :width="embedImage.width"
          :height="embedImage.height"
          alt="Image."
          class="embed__image"
          data-testid="embed__image"
        >
      </div>
    </template>
  </Component>
</template>

<script>
export default {
  name: 'EmbedSlice',
  props: {
    slice: {
      type: Object,
      default: () => ({}),
    },
  },

  data() {
    return {
      item: this.slice?.items[0],
      embedImage: {
        url: `${this.slice?.items[0]?.embed?.thumbnail_url}?w=400&h=218`,
        width: '200',
        height: '109',
      },

      embedFieldData: {},
    }
  },

  computed: {
    isTag() {
      if (this.item?.embed?.type === 'link') {
        return {
          is: 'a',
          target: '_blank',
          href: this.item?.embed?.url,
          class: `embed__${this.item?.embed?.type}`,
        }
      }
      return {
        is: 'div',
        class: `embed__${this.item?.embed?.type}`,
      }
    },
  },

  mounted() {
    this.onResize()
    window.addEventListener('resize', this.onResize)
  },

  beforeDestroy() {
    window.removeEventListener('resize', this.onResize)
  },

  created() {
    if (!this.item) return

    const {
      embed: {
        title: rawTitle,
        html: rawHtml,
        type: embedType,
      },
      embed_title: embedTitle,
      embed_description: embedDescription,
    } = this.item

    if (rawHtml) {
      const matchDescription = rawHtml.match('<p>(.*?)</p>')
      const description = matchDescription ? matchDescription[1] || '' : ''

      const html = `
        <div class="embed__title">${embedTitle || rawTitle.split(' | ')[0]}</div>
        <p class="embed__description">${embedDescription || description}</p>
      `

      this.embedFieldData = {
        ...this.slice?.embed,
        html,
      }
    }

    if (embedType === 'video') {
      const html = rawHtml
        .replace(/height="[0-9]*"/, 'height="500"')
        .replace(/width="[0-9]*"/, 'width="100%"')

      this.embedFieldData = {
        ...this.slice?.embed,
        html,
      }
    }
  },

  methods: {
    onResize() {
      if (this.item && this.item?.embed?.type === 'link') {
        const changeEmbedImage = (url, width, height) => {
          this.embedImage.url = url
          this.embedImage.width = width
          this.embedImage.height = height
        }
        if (window.innerWidth < 1024) changeEmbedImage(`${this.item?.embed?.thumbnail_url}?w=1164&h=632`, '582', '316')
        else changeEmbedImage(`${this.item?.embed?.thumbnail_url}?w=400&h=218`, '200', '109')
      }
    },
  },
}
</script>

<style lang="scss" scoped>
.embed {
  /deep/ iframe {
    display: block;
    max-width: 100%;
  }

  &__link,
  &__video {
    margin: 25px 0;
  }

  &__link {
    display: flex;
    align-items: center;
    justify-content: space-between;
    border: 1px solid $border-color--silver;
    text-decoration: none;
    @media screen and (max-width: 1024px) {
      flex-wrap: wrap;
    }

    &-content {
      width: 100%;
      padding: 24px;
      padding-right: 0;
      @media screen and (max-width: 1024px) {
        order: 2;
        width: 100%;
        padding-right: 24px;
      }
    }
  }

  /deep/ &__title,
  /deep/ &__description {
    letter-spacing: -0.02em;
    line-height: 129%;
  }

  /deep/ &__title {
    @include font('Poppins', 21px, 700);
    line-height: 27.34px;
    color: $text-color--black-oil;
    margin: 0;
    margin-bottom: 6px;
  }

  /deep/ &__description {
    @include font('Inter', 13px, 400);
    color: $text-color--grey-pale;
    margin: 0;
  }

  /deep/ &__image {
    display: block;
    width: 200px;
    min-width: 200px;
    height: auto;
    object-fit: cover;
    @media screen and (max-width: 1024px) {
      width: 100%;
      min-width: auto;
    }

    &-wrapper {
      width: 33%;
      display: flex;
      align-items: center;
      padding: 24px;
      @media screen and (max-width: 1024px) {
        order: 1;
        width: 100%;
      }
    }
  }
}
</style>
