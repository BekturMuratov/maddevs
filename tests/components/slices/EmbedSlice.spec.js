// eslint-disalbe quotes
import { render } from '@testing-library/vue'
import EmbedSlice from '@/components/slices/EmbedSlice'

describe('EmbedSlice component', () => {
  const props = {
    slice: {
      items: [
        {
          embed: {
            title: 'Title',
            url: 'https://test',
            type: 'video',
            thumbnail_url: '',
            html:
              '<div><img src="" height="300" width="300"/></div><h1>hello</h1><a href="http://maddevs.io">Hi</a> </div>',
          },
          embed_title: 'Embed title',
        },
      ],
      primary: {
        code: [
          {
            spans: [],
            text:
              'getAllHeadings() {\n  this.document.body.forEach(listItem => {\n    if(listItem.primary.text !== undefined) {\n      if(listItem.primary.text[0].type === \'heading1\') {\n        this.headingsList.push({\n          textContent: listItem.primary.text[0].text,\n          headingName: listItem.primary.text[0].text.toLowerCase().replace(/s/g , \'-\')\n        });\n      }\n    }\n  });',
            type: 'paragraph',
          },
        ],
        language: 'javascript',
      },
      slice_label: null,
      slice_type: 'codeblock',
    },
  }
  it('should render correctly', () => {
    const { container } = render(EmbedSlice, {
      props,
      stubs: ['PrismicEmbed'],
      mocks: {
        $prismic: {
          asHtml: name => name,
        },
      },
    })

    expect(container).toMatchSnapshot()
  })

  it('must replace tags', async () => {
    const { container } = await render(EmbedSlice, {
      props: {
        slice: {
          items: [],
          primary: {
            code: [],
            language: 'javascript',
          },
          slice_label: null,
          slice_type: 'codeblock',
        },
      },
      stubs: ['PrismicEmbed'],
    })

    expect(container).toMatchSnapshot()
  })
})
