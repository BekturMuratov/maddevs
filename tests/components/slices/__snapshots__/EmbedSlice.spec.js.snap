// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmbedSlice component must replace tags 1`] = `
<div>
  <!---->
</div>
`;

exports[`EmbedSlice component should render correctly 1`] = `
<div>
  <div
    class="embed__video"
  >
    <prismicembed-stub
      class="embed__video-content"
      field="[object Object]"
    />
     
    <!---->
  </div>
</div>
`;
